File: Algorithm/contains_duplicate_iii.py
import random
import collections
class Solution:
    def containsNearbyAlmostDuplicate(self, nums, k, t):
        dic = collections.OrderedDict()          
        for val in nums:
            key  = val / t if t else val

            if dic.get(key) or (dic.get(key-1) and (dic.get(key-1) - val<= t)) or  (dic.get(key+1) and (dic.get(key+1) - val <= t)):
                return True
            print dic
            if len(dic) == k:
                dic.popitem(False)
            dic[key] = val
            print dic
        return False
    def containsNearbyAlmostDuplicate1(self, nums, k, t):
        dic = collections.OrderedDict()          
        for val in nums:
            key  = val  if not t else val/t 
            for pre in ( dic.get(key-1), dic.get(key),dic.get(key+1)):
                print pre, val
                if pre and abs(pre - val) <= t:
                    return True
            if len(dic) == k:
                dic.popitem(False)
            dic[key] = val
        return False            
                
if __name__ == '__main__':
    sol = Solution()
    nums = [random.randint(1, 20) for i in range(10)]
    k = random.randint(1, 10)
    t = random.randint(1, 10)
    nums = [2,2]
    k = 3
    t = 0
    nums = [1,3,1]
    k = 1
    t = 1
    nums = [3,6,0,2]
    k = 2
    t = 2
    print nums, k, t
    print sol.containsNearbyAlmostDuplicate(nums, k, t)
    print sol.containsNearbyAlmostDuplicate1(nums, k, t)

File: Algorithm/CPP/contains_duplicate_iii.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
# include <string>
# include <unordered_map>
# include <cctype>
# include <algorithm>
# include <time.h>
using namespace std;

class Solution {
public:
	bool containsNearbyAlmostDuplicate(vector<int>& nums, int k, int t) {
		if (k < 1 || t < 0)return false;
		unordered_map<int, int> dic;
		int key;
		for (int i = 0; i < nums.size(); i++)
		{
		
			if (t>1)
			{
				key = nums[i] / t;
				if (nums[i] < 0 && nums[i]%t) key--;
			}
			else key = nums[i];
			if ( dic.find(key) != dic.end() ||
				(dic.find(key - 1) != dic.end() && abs(dic[key - 1] - nums[i]) <= t) ||
				(dic.find(key + 1) != dic.end() && abs(dic[key + 1] - nums[i]) <= t) )
				return true;
			if (dic.size() == k) dic.erase(dic.begin());
			dic[key] = nums[i];
		}
		return false;
	}
};

int main()
{
	Solution sol;
	vector<int> nums = { 2, 0, -2, 2  };
	cout << sol.containsNearbyAlmostDuplicate(nums, 2, 1);

	cout << endl;
	//for (auto k :nums )cout << k << " ";
	system("pause");
	return 0;
}

